### DISCUSSION TITLE
cupy instead of dask?

### DISCUSSION THREAD

**Discussion by rajeeja** [UXARRAY/uxarray#950] [General]
JIT and dask have worked fine for us so far. During my meetings at SC 2022 @anissa111  and some recent developments it has come to light that cupy might be a good long term alternative to dask.

It would be good to have a small prototype and discussion on if this options should be explored.

**Comment by erogluorhan** [comment]
So GPU instead of CPU parallelization, or compatibility with both maybe? I am happy to discuss this further.

**Comment by dcherian** [comment]
It's feasible (and fun!) to do both.

- numba can compile to a CUDA target, so the existing kernels could be generalized
- dask can wrap cupy, just like it wraps numpy;

I think the only real annoyance is manually switching between the CUDA kernel and CPU kernel depending on array type

Datashader has been doing this for a while: https://anaconda.org/jonmmease/datashader-rapids-histogram-example/notebook

Another example: https://examples.dask.org/applications/stencils-with-numba.html

**Comment by rljacob** [comment]
It looks like cupy is for on-node parallelim within a gpu while dask is task-parallelism across cpu nodes).  So one is not a replacement for the other.  What am I missing?

Does cupy only target NVIDIA Gpus?   There are 2 others that DOE at least needs to support.

**Comment by rajeeja** [comment]
It'd be interesting to revisit this now. It has been 2 years. Just putting this comment as a reminder for me to do some more testing along these lines.
